ambry-network/src/test/java/com/github/ambry/network/http2/AmbrySendToHttp2AdaptorTest.java

```diff
@@ -54,31 +55,44 @@ public void after() {
    */
   @Test
   public void testServerWrite() throws Exception {
-    EmbeddedChannel channel = newChannel();
-    // Test server writing the Send
-    Send send = new SendWithContent();
+    int maxFrameSize = 2001;
+    EmbeddedChannel channel = new EmbeddedChannel(new AmbrySendToHttp2Adaptor(true, maxFrameSize));
+    int contentSize = 7000;
+    byte[] byteArray = new byte[contentSize];
+    new Random().nextBytes(byteArray);
+    ByteBuf content = UnpooledByteBufAllocator.DEFAULT.heapBuffer(contentSize).writeBytes(byteArray);
+    // Retain the ByteBuf for data comparison
+    content.retain();
+    Send send = new MockSend(content);
     channel.writeOutbound(send);
 
     DefaultHttp2HeadersFrame header = channel.readOutbound();
     Assert.assertNotNull(header.headers());
     Assert.assertEquals(header.headers().status().toString(), "200");
 
+    byte[] resultArray = new byte[contentSize];
+    int i;
+    for (i = 0; i < contentSize / maxFrameSize; i++) {
+      DefaultHttp2DataFrame data = channel.readOutbound();
+      data.content().readBytes(resultArray, i * maxFrameSize, maxFrameSize);
+      data.content().release();
+    }
+    // The last frame
     DefaultHttp2DataFrame data = channel.readOutbound();
+    data.content().readBytes(resultArray, i * maxFrameSize, data.content().readableBytes());
     data.content().release();
-  }
-
-  /**
-   * @return a new {@link EmbeddedChannel} with {@link AmbrySendToHttp2Adaptor} added to the pipeline.
-   */
-  static EmbeddedChannel newChannel() {
-    return new EmbeddedChannel(new AmbrySendToHttp2Adaptor(true));
+    Assert.assertArrayEquals(byteArray, resultArray);
```

Comment:
shouldn't you call `content.release()` after you finish comparison given that you already call content.retain() before?

Reply:
Didn't get you. Data was first read into the resultArray, then content is released immediately.

Reply:
Fixed. (Looks like only Pool ByteBuf needs release. )